src/shell.zig

```diff
@@ -31,13 +33,19 @@ arena: std.heap.ArenaAllocator,
 
 /// Root directory of this repository.
 ///
-/// Prefer this to `std.fs.cwd()` if possible.
-///
 /// This is initialized when a shell is created. It would be more flexible to lazily initialize this
 /// on the first access, but, given that we always use `Shell` in the context of our repository,
 /// eager initialization is more ergonomic.
 project_root: std.fs.Dir,
 
+/// Shell's logical cwd which is used for all functions in this file. It might be different from
+/// `std.fs.cwd()` and is set to `project_root` on init.
+cwd: std.fs.Dir,
+
+// Stack of working directories backing pushd/popd.
+cwd_stack: [cwd_stack_max]std.fs.Dir,
```

Comment:
Open-coded bounded array here:

* std.BoundedArray is banned in favor of stdx.BoundedArray
* but, as we use shell.zig in build.zig, we don't want  to drag any deps here
* and, also, std.BoundedArray doesn't really by us all that much.